### PR TITLE
Add Optional Language Variable to WelcomeTitle Component [New PR]

### PR DESCRIPTION
**Pull request by grvpandey11**
## Hey, I just made a Pull Request!

This Pull Request adds an optional language variable to the 'WlcomeTitle' component, allowing the greeting to be customized based on the selected language.
The 'getTimeBasesGreeting' function has been updated to accept the language parameter and fetch the respective greeting from the JSON file.

Old PR for reference: https://github.com/backstage/backstage/pull/18519
#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by grvpandey11** [comment]
@awanlin Can you please review.
I have made the changes which you suggested along with Changeset and api-report.md file.

**Comment by grvpandey11** [comment]
This PR will fix https://github.com/backstage/backstage/issues/18520

File: plugins/home/api-report.md [hunk 1]
```diff
@@ -44,7 +44,7 @@ export type ClockConfig = {
 export const ComponentAccordion: (props: {
   title: string;
   expanded?: boolean | undefined;
-  Content: () => JSX.Element;
+  Content: () => JSX.Element /** @public */;
```
Comment (awanlin) [inline plugins/home/api-report.md:47]: Hmm, I'm not sure why this has been changed, does not seem related 🤔

File: plugins/home/src/homePageComponents/WelcomeTitle/WelcomeTitle.test.tsx [hunk 2]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright 2021 The Backstage Authors
+ * Copyright 2023 The Backstage Authors
```
Comment (awanlin) [inline plugins/home/src/homePageComponents/WelcomeTitle/WelcomeTitle.test.tsx:2]: You can leave this, I think the general pattern is that it represents when the code was first written. But let's let the maintainers confirm this one

File: .changeset/quiet-starfishes-kick.md [hunk 3]
```diff
@@ -0,0 +1,5 @@
+---
+'@backstage/plugin-home': patch
+---
+
+The `getTimeBasedGreeting` function has been updated to support an optional language parameter. Now, callers of the function can provide a language as input to receive a greeting in that specific language. If no language is provided, the function will automatically select a random greeting based on the current time.
```
Comment (awanlin) [inline .changeset/quiet-starfishes-kick.md:5]: This is good but can we include an example, something as simple as this would work: `<WelcomeTitle language={['English' 'Spanish']} />`

**Review by awanlin — CHANGES_REQUESTED** [review]
This looks pretty solid to me, thanks again @grvpandey11 🚀 

Left a couple of small items

File: plugins/home/api-report.md (hunk 1 above)
Comment (awanlin) [reply plugins/home/api-report.md:47]: I have a feeling this is related to the build failures as well

Comment (grvpandey11) [reply plugins/home/api-report.md]: Fixed

File: plugins/home/src/homePageComponents/WelcomeTitle/WelcomeTitle.test.tsx (hunk 2 above)
Comment (grvpandey11) [reply plugins/home/src/homePageComponents/WelcomeTitle/WelcomeTitle.test.tsx]: fixed

File: .changeset/quiet-starfishes-kick.md (hunk 3 above)
Comment (grvpandey11) [reply .changeset/quiet-starfishes-kick.md]: fixed

**Comment by awanlin** [comment]
Hi @camilaibs, @emmaindal, could you take a look at this PR please? I think its in a stage where you feedback and approval would be great. Thanks in advance!

File: .changeset/quiet-starfishes-kick.md [hunk 4]
```diff
@@ -0,0 +1,5 @@
+---
+'@backstage/plugin-home': patch
+---
+
+Now, user can provide a language (optional) as input to receive a greeting in that specific language. Example: <WelcomeTitle language={['English' 'Spanish']} />
```
Comment (awanlin) [inline .changeset/quiet-starfishes-kick.md:5]: Small optional change: maybe put backticks (`) around the example so it shows as a code block?

**Review by awanlin — APPROVED** [review]
Alright, this is great, builds are failing due to an issue not related to the code change so don't worry about them

Many thanks again! 👍

**Comment by grvpandey11** [comment]
@awanlin I think all the checks passed. I am not sure about the build which is failing is because of my changes, but I can see the same is failing in others PR also.  So, I think I am good with the changes. Hope to see PR merge soon

File: .changeset/quiet-starfishes-kick.md (hunk 4 above)
Comment (grvpandey11) [reply .changeset/quiet-starfishes-kick.md]: fixed

**Comment by awanlin** [comment]
@grvpandey11, we need the maintainers to approve as well so that's the next step. I suspect that we should get that pretty easily once they take a look. Keep in mind that it's vacation time so we might run into delays due to that.

**Comment by awanlin** [comment]
Hi @tikabom, would you be able to look at this PR, please? Sorry to ping you directly but would be great to get this into the release tomorrow 🙏

**Comment by camilaibs** [comment]
Hi there, thanks in advance for the contribution! I'm starting to review (sorry for the late reply, I was on summer vacation)...

File: plugins/home/src/homePageComponents/WelcomeTitle/timeUtil.ts [hunk 5]
```diff
@@ -22,7 +22,10 @@ import goodEvening from './locales/goodEvening.locales.json';
 // every time the user navigates.
 const greetingRandomSeed = Math.floor(Math.random() * 1000000);
 
-export function getTimeBasedGreeting(): { language: string; greeting: string } {
+export function getTimeBasedGreeting(language?: string[] | undefined): {
+  language: string;
```
Comment (camilaibs) [inline plugins/home/src/homePageComponents/WelcomeTitle/timeUtil.ts:26]: Would it be possible to export an enum with the existing keys in the locale files? If that's possible, we could use it as a language type to avoid string typos... That way people would be able to define the language array with just the keys that exist in the locale files. Another cool ability would be to accept a single string to get greetings always in the same language. Here is an example:
```tsx
import { WelcomeTitle, welcomeTitleLanguages } from '@backstage/plugin-home';

// Multiple languages with an array
<WelcomeTitle language={[welcomeTitleLanguages.English, welcomeTitleLanguages.Swedish]} />

// Single language with a string
<WelcomeTitle language={welcomeTitleLanguages.English} />

File: plugins/home/src/homePageComponents/WelcomeTitle/WelcomeTitle.test.tsx [hunk 6]
```diff
@@ -30,3 +31,15 @@ describe('<WelcomeTitle>', () => {
     expect(getByText(/Get some rest, Guest/)).toBeInTheDocument();
   });
 });
+
+test('should greet user with a single language', async () => {
```
Comment (camilaibs) [inline plugins/home/src/homePageComponents/WelcomeTitle/WelcomeTitle.test.tsx:35]: Thanks for increasing the code coverage 🙌🏻

File: plugins/home/src/homePageComponents/WelcomeTitle/timeUtil.ts [hunk 7]
```diff
@@ -22,7 +22,10 @@ import goodEvening from './locales/goodEvening.locales.json';
 // every time the user navigates.
 const greetingRandomSeed = Math.floor(Math.random() * 1000000);
 
-export function getTimeBasedGreeting(): { language: string; greeting: string } {
+export function getTimeBasedGreeting(language?: string[] | undefined): {
```
[the remark this answers is not in the document]
Comment (grvpandey11) [reply plugins/home/src/homePageComponents/WelcomeTitle/timeUtil.ts]: Fixed

File: plugins/home/src/homePageComponents/WelcomeTitle/timeUtil.ts (hunk 5 above)
Comment (grvpandey11) [reply plugins/home/src/homePageComponents/WelcomeTitle/timeUtil.ts:26]: Initially I have setup input for single language but than I received suggestion to convert it to array where we can input multiple as well as single language.

I am having difficulty in converting it to enum at the moment. will it be blocker in merging the PR?
@camilaibs

File: plugins/home/src/homePageComponents/WelcomeTitle/WelcomeTitle.test.tsx (hunk 6 above)
Comment (grvpandey11) [reply plugins/home/src/homePageComponents/WelcomeTitle/WelcomeTitle.test.tsx:35]: Anything to do here?

File: plugins/home/src/homePageComponents/WelcomeTitle/timeUtil.ts [hunk 8]
```diff
@@ -32,12 +35,26 @@ export function getTimeBasedGreeting(): { language: string; greeting: string } {
...
   const timeOfDay = (hour: number): { [language: string]: string } => {
     if (hour < 12) return goodMorning;
     if (hour < 17) return goodAfternoon;
     return goodEvening;
   };
+
   const greetings = timeOfDay(currentHour);
+
+  if (Array.isArray(language) && language.length > 0) {
+    const validLanguages = language.filter(lang => lang && greetings[lang]);
```
[the remark this answers is not in the document]
Comment (grvpandey11) [reply plugins/home/src/homePageComponents/WelcomeTitle/timeUtil.ts:48]: Need help here. How can I introduce log here, when I do "console.log" it prints the error in browser. But I guess the expectation is to have it with application logs.
@awanlin @camilaibs

File: plugins/home/src/homePageComponents/WelcomeTitle/WelcomeTitle.test.tsx (hunk 6 above)
Comment (camilaibs) [reply plugins/home/src/homePageComponents/WelcomeTitle/WelcomeTitle.test.tsx:35]: No, just sharing appreciation, thank you again 🙂

File: plugins/home/src/homePageComponents/WelcomeTitle/timeUtil.ts (hunk 5 above)
Comment (camilaibs) [reply plugins/home/src/homePageComponents/WelcomeTitle/timeUtil.ts:26]: No blocking, it was just a suggestion if possible.
If you later find an idea on how to do it, you are more than welcome to contribute to us again in a separate PR.

File: plugins/home/src/homePageComponents/WelcomeTitle/timeUtil.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (camilaibs) [reply plugins/home/src/homePageComponents/WelcomeTitle/timeUtil.ts:48]: I was thinking a simple console is enough, in case developers inspect the page to see what's going on, but a totally optional suggestion, we can merge without it 🙂

**Review by camilaibs — APPROVED** [review]
LGTM, thank you again 🎉

**Comment by grvpandey11** [comment]
@camilaibs Thanks for approving it.
I think I will not have permission to merge the Pull Request.
Can you help me with it.

**Comment by awanlin** [comment]
@camilaibs @benjdlambert any chance one of you can merge this, please? I know I can as an Org Member but it's not something I've done in the past so playing it safe, would be super nice to get this into the next release!

**Comment by awanlin** [comment]
Thanks @camilaibs 🚀

**Comment by grvpandey11** [comment]
Thanks @camilaibs @awanlin for the review.
